### ISSUE TITLE
AudioStreamImportSettingsDialog: loop offset field only allows values with up to three decimals, causing a loop point imprecision that sometimes leads to an audible pop when looping

### ISSUE THREAD

**Issue by RustyRoboticsBV** [godotengine/godot#87427]
### Tested versions

Reproducable in: v4.1.1.stable.mono.official [bd6af8e0e], v4.3.dev2.official [352434668]

### System information

Godot v4.3.dev2 - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1650 (NVIDIA; 30.0.14.7219) - AMD Ryzen 7 4800H with Radeon Graphics (16 Threads)

### Issue description

Some of my ogg music files that use the loop offset variable have audible pops whenever they loop. I've done some testing, and the issue appears to be the following: the ogg importer only lets you set loop time offsets in up to three decimals of precision. With the default sample rate of 44,100 Hz, this creates a potential imprecision of 0.0009999999 * 44,100 ≈ 44 samples during a loop.

For example, I have an ogg file where the loop point needs to be at sample 651,323 for a seamless loop. With a sample rate of 44,100 Hz, that gets me a time offset value of 14.7692307692 seconds.
Unfortunately, the import window only allows a time offset of either 14.769s (sample 651,312) or 14.770s (sample 651,357), missing the mark by either 11 or 34 samples respectively.

The result is an audible pop in some songs, which I've confirmed by comparing a recording of the loop to the original file in Audacity. If I manually set the loop point at runtime through AudioStream.Set("loop_offset", <higher_precision_value>), higher-precision values *are* allowed and the issue disappears. All songs that I've tested seem to loop perfectly after doing this.

Editing the "default float step" editor setting did not fix the problem. Manually editing the .import file in notepad doesn't work either: the value still gets imported with the loop offset rounded to three decimals.

This problem was already mentioned by @strellydev in #80452 and #64775, but I felt it was worth creating a more targeted thread for this issue, as I suspect it's more a resource loading problem than a problem in the ogg implementation itself.

I'm going to investigate the issue and post my findings. If the problem is what I think it is, I propose increasing the precision of the loop offset field to at least five decimals.

### Steps to reproduce

1. Add an ogg file to project.
2. Enable looping.
3. Enter a value in loop offset field. The value gets rounded to three decimals.
4. Depending on the music file, audio pops may occur on loop.

### Minimal reproduction project (MRP)

N/A

**Comment by bs-mwoerner** [comment]
My practical advice and what I usually do is: Edit the audio files so that the loop points are at whole second offsets, although that's indeed a bit silly. If this is no longer an issue with the internal implementation after #80452, then a true fix could be as simple as allowing the loop point to be given in samples instead of seconds and then just convert that to the internal 64 bit floating point time value.

Ah, for WAV, the loop point is (only) specified in samples, so there's some inconsistency here anyway.

**Comment by Bermast** [comment]
Update: The culprit appears to be line 546 of `audio_stream_import_settings.cpp`: `loop_offset->set_step(0.001);`

Unfortunately, after changing it to a more precise step, loop offset values _still_ get rounded. However, the decimal position that values get rounded on is now inconsistent, and seems to depend on the number of digits before the decimal point (higher numbers can have less decimals, lower numbers can have more). Smells like casting-induced floating-point imprecision to me, but it might also be that values get rounded somewhere else.

> a true fix could be as simple as allowing the loop point to be given in samples instead of seconds and then just convert that to the internal 64 bit floating point time value.
> Ah, for WAV, the loop point is (only) specified in samples, so there's some inconsistency here anyway.

I like this suggestion, though I'm a little concerned about breaking existing projects with this fix. Still, I'm going to try and implement it, see if I can get things working.

Perhaps I could look into implementing a toggle of some kind, which would allow users to switch between seconds and sample values? This would preserve existing projects, while still giving more precision when needed. On the plus side, I think it could probably be implemented without even touching `AudioStreamOggVorbis`, so the importer would handle all the conversions.

**Comment by RustyRoboticsBV** [comment]
Update: The culprit appears to be line 546 of audio_stream_import_settings.cpp: loop_offset->set_step(0.001);

Unfortunately, after changing it to a more precise step, loop offset values still get rounded. However, the decimal position that values get rounded on is now inconsistent, and seems to depend on the number of digits before the decimal point (higher numbers can have less decimals, lower numbers can have more). Smells like casting-induced floating-point imprecision to me, but it might also be that values get rounded somewhere else.

> a true fix could be as simple as allowing the loop point to be given in samples instead of seconds and then just convert that to the internal 64 bit floating point time value.
> Ah, for WAV, the loop point is (only) specified in samples, so there's some inconsistency here anyway.

I like this suggestion, though I'm a little concerned about breaking existing projects with this fix. Still, I'm going to try and implement it, see if I can get things working.

Perhaps I could look into implementing a toggle of some kind, which would allow users to switch between seconds and sample values? This would preserve existing projects, while still giving more precision when needed. On the plus side, I think it could probably be implemented without even touching AudioStreamOggVorbis, so the importer would handle all the conversions.

**Comment by bs-mwoerner** [comment]
> I like this suggestion, though I'm a little concerned about breaking existing projects with this fix. Still, I'm going to try and implement it, see if I can get things working.

Yes, I was thinking about having the samples input as an additional UI option without changing anything internally, so that compatibility is not an issue. One would probably have to make sure that a seconds value is only applied if it is significantly different from the current internal value, or it would be easy to inadvertently overwrite the current value with a rounded version just by switching formats.

**Comment by RustyRoboticsBV** [comment]
Update: I think I've gotten things working! The imprecision-induced looping pop seems to be fixed now, for both ogg and mp3 (as the importer dialog is shared between the two).

I'm going to try and figure out how to make a pull request now.

**Comment by MJacred** [comment]
@RustyRoboticsBV: could you check https://github.com/godotengine/godot/pull/87860 to see if it fits your needs, please?

It would have have way less changes compared to https://github.com/godotengine/godot/pull/87554, which will most likely take quite some discussion with an alternate approach in the end (is what my gut tells me)

**Comment by RustyRoboticsBV** [comment]
> @RustyRoboticsBV: could you check #87860 to see if it fits your needs, please?
Building your code as we speak!

**Comment by DeeJayLSP** [comment]
IMO, the best workaround right now would be to add the loop offset with multiple decimals directly into the corresponding `.import` file. This won't be modified by the editor unless you touch the value there.

**Comment by RustyRoboticsBV** [comment]
> IMO, the best workaround right now would be to add the loop offset with multiple decimals directly into the corresponding `.import` file. This won't be modified by the editor unless you touch the value there.

I agree, and I actually looked into implementing something like this a while back. The issue that I ran into was that floating-point values are hard-coded to be truncated in the function that handles serialization (see [here](https://github.com/godotengine/godot/pull/87554#issuecomment-1924092560)). If I remember correctly (it's been a while) the relevant code had some preprocessor directives for different operating systems, complicating matters a bit.

There was also some [discussion](https://github.com/godotengine/godot/pull/87554#issuecomment-1911092439) on that page about floating-point imprecision, and whether or not integers (representing a sample index) should be used internally to encode the loop point. The idea here being that this would guarantee sample-level precision for longer songs. Perhaps more discussion is needed here.

**Comment by Akosmo** [comment]
I think I agree that `loop_offset` should be in samples rather than seconds. It's not as beginner friendly, but I think it's better to calculate the right value and avoid pops, than being stuck with pops. I'm not sure why this isn't the case already. Maybe because aside from being more beginner friendly, maybe it's easier to work with time rather than samples in MP3 and OggVorbis...?

**Comment by Jaytheway** [comment]
When authoring seamlessly looping sounds, usually you either have to match your exports to have zero-crossing at rounded fraction of a second bound, or work with sample values. The latter one is rather straightforward. All the DAWs provide sample timeline display and precise export ranges.

Using stable sample type internally and changing only the way they are displayed in the editor is a common solution, and, in my opinion, would be reasonable.
Which is more-or-less how it is right now (ogg and mp3 streams convert to samples anyway); only there's no switch to change the displayed and editable units.
